* Mathematicians do not prove every theorem from scratch. Instead, they build their proofs on the truths their predecessors have already established
* In the same way, it is vanishingly rare for someone to write all of a program alone
* It is much more common — and productive — to make use of the millions of lines of code that other programmers have written before

* A $\rm\color{orange}{module}$ is a collection of variables and functions that are grouped together in a single file
* The variables and functions in a module are usually related to one another in some way
    * For example, module math contains the variable pi and mathematical functions such as cos (cosine) and sqrt (square root)
* This lecture shows how to use some of the hundreds of modules that come with Python

### Importing Modules<br>
* To gain access to the variables and functions from a module, we have to import it
* To tell Python that you want to use functions in module math, for example, we use this $\rm\color{orange}{import}$ statement

In [2]:
import math

* Importing a module creates a new variable with that name
* That variable refers to an object whose type is module

In [3]:
print(type(math))

<class 'module'>


* Once you have imported a module, you can use built-in function help to see what it contains

In [4]:
help(math)

Help on built-in module math:

NAME
    math

DESCRIPTION
    This module provides access to the mathematical functions
    defined by the C standard.

FUNCTIONS
    acos(x, /)
        Return the arc cosine (measured in radians) of x.
        
        The result is between 0 and pi.
    
    acosh(x, /)
        Return the inverse hyperbolic cosine of x.
    
    asin(x, /)
        Return the arc sine (measured in radians) of x.
        
        The result is between -pi/2 and pi/2.
    
    asinh(x, /)
        Return the inverse hyperbolic sine of x.
    
    atan(x, /)
        Return the arc tangent (measured in radians) of x.
        
        The result is between -pi/2 and pi/2.
    
    atan2(y, x, /)
        Return the arc tangent (measured in radians) of y/x.
        
        Unlike atan(y/x), the signs of both x and y are considered.
    
    atanh(x, /)
        Return the inverse hyperbolic tangent of x.
    
    cbrt(x, /)
        Return the cube root of x.
    
    ceil(x, /)

* The statement import math creates a variable called math that refers to a module
object
* In that object are all the names defined in that module
* Each of them refers to a function object
---
![MathModule](lec05-01.jpg)

* When we try to calculate a square root, though, we get an error telling us that Python is still unable to find function $sqrt$

In [5]:
print(sqrt(9))

NameError: name 'sqrt' is not defined

* The solution is to tell Python explicitly to look for the function in module math by combining the module’s name with the function’s name using a $\rm\color{magenta}{dot}$

In [5]:
print(math.sqrt(9))

3.0


* The dot is an $\rm\color{magenta}{operator}$, just like $+$ and $**$ are operators
* Its meaning is "look up the object that the variable to the left of the dot refers to and, in that object, find the name that occurs to the right of the dot"
* In $math.sqrt(9)$, Python finds math in the current namespace, looks up the module object that math refers to, finds function sqrt inside that module, and then executes the function call

* Modules can contain more than just functions
* Module math, for example, also defines some variables like $\rm\color{orange}{pi}$
* Once the module has been imported, you can use these variables like any others

In [6]:
import math

print(math.pi)
radius = 5
print('Area is', math.pi * radius ** 2)

3.141592653589793
Area is 78.53981633974483


* We can even assign to variables imported from modules

In [7]:
import math

math.pi = 3
radius = 5
print('Area is', math.pi * radius ** 2)

Area is 75


* $\rm\color{red}{DON'T\space DO\space THIS!}$ Changing the value of $\pi$ is not a good idea
* In fact, it is such a bad idea that many languages allow programmers to define unchangeable constants as well as variables
* As the name suggests, the value of a constant cannot be changed after it has been defined
    * $\pi$ is always 3.14159 and a little bit
    * SECONDS_PER_DAY is always 86,400
* The fact that Python does not allow programmers to "freeze" values like this is one of the language’s few significant
flaws

* Combining the module’s name with the names of the things it contains is safe, but it is not always convenient
* For this reason, Python lets you specify exactly what you want to import from a module

In [4]:
from math import sqrt, pi

print(sqrt(9))
radius = 5
print('Circumference is', 2 * pi * radius)

3.0
Circumference is 31.41592653589793


* This does not introduce a variable called $math$
* Instead, it creates function $sqrt$ and variable $pi$ in the current namespace, as if you had typed the function definition and variable assignment yourself
* Restart the kernel and try the following

In [5]:
from math import sqrt, pi

print(math.sqrt(9))


NameError: name 'math' is not defined

In [3]:
print(sqrt(9))

3.0


* Here, we do not have a variable called $math$
* Instead, we imported variables $sqrt$ and $pi$ directly into the current namespace
---
![FromImport](lec05-02.jpg)

* This can lead to problems when different modules provide functions that have the same name
* If you import a function called $\rm\color{aqua}{spell}$ from a module called $\rm\color{aqua}{magic}$ and then you import another function called $\rm\color{lime}{spell}$ from the $\rm\color{lime}{grammar}$ module, the second replaces the first
* It is exactly like assigning one value to a variable and then assigning another value: The most recent assignment or import wins

* This is why it is usually $\rm\color{magenta}{NOT}$ a good idea to use $\rm\color{magenta}{import\space *}$, which brings in everything from the module at once
* Although $\rm\color{magenta}{import\space *}$ saves some typing, you run the risk of your program accessing the incorrect function and not working properly

In [ ]:
from math import *

print(sqrt(8))

* The standard Python library contains several hundred modules to do everything from figuring out what day of the week it is to fetching data from a website
* The full list is online at [here](https://docs.python.org/3/py-modindex.html)
* Knowing how to use the library well is one of the things that distinguishes good programmers from poor ones